### PR TITLE
Switch to using Codecov GitHub Action.

### PR DESCRIPTION
**Pull request by nmoinvaz**
This uses codecov's GitHub Actions action. 
https://github.com/codecov/codecov-action

Which also uses their new uploader.

### CODE REVIEW

**Comment by nmoinvaz** [comment]
@Dead2 @iii-i Is it possible to install these packages on `z15` runner?
```
sudo apt-get install libxml2-dev libxslt-dev python-dev
```
https://stackoverflow.com/questions/5178416/libxml-install-error-using-pip

**Comment by nmoinvaz** [comment]
Fixed pigz workflow.

**Comment by iii-i** [comment]
I've installed `libxml2-devel libxslt-devel python3-devel` (the builder is running RHEL).

**Comment by nmoinvaz** [comment]
It looks like this command is still failing:
`python3 -u -m pip install --user gcovr`

**Comment by iii-i** [comment]
Argh, of course - the build is running inside a Docker container. I've updated the image, please try again. If it works, feel free to cherry-pick https://github.com/iii-i/zlib-ng/commit/f26c34768802eb1f1fb614a50233a6ae1e771bce into your PR.

**Comment by nmoinvaz** [comment]
Cherry-picked. Thanks!

**Comment by nmoinvaz** [comment]
@Dead2 what is the plan for buildkite?

**Comment by Dead2** [comment]
This does not seem to work right.
Codecov.io shows "No report uploaded yet" for this PR, other PRs (like #1159) work fine.

Buildkite.. not decided yet, but was considering using it for automated benchmarking.

**Comment by Dead2** [comment]
I wonder why this PRs coverage does not appear in codecov.
It seems like the token is sent correctly, and it seems like it finds all the coverage data, and it probably gets uploaded. But codecov .io sees no reports..

```
Processing file: /home/runner/work/zlib-ng/zlib-ng/CMakeFiles/zlib.dir/insert_string.c.gcda
Running gcov: 'gcov /home/runner/work/zlib-ng/zlib-ng/CMakeFiles/zlib.dir/insert_string.c.gcda --branch-counts --branch-probabilities --preserve-paths --object-directory /home/runner/work/zlib-ng/zlib-ng/CMakeFiles/zlib.dir' in '/tmp/tmpx5lhk_69'
Finding source file corresponding to a gcov data file
  currdir      /home/runner/work/zlib-ng/zlib-ng
  gcov_fname   /tmp/tmpx5lhk_69/#home#runner#work#zlib-ng#zlib-ng#insert_string_tpl.h.gcov
               ['        -', '    0', 'Source', '/home/runner/work/zlib-ng/zlib-ng/insert_string_tpl.h\n']
  source_fname /home/runner/work/zlib-ng/zlib-ng/CMakeFiles/zlib.dir/insert_string.c.gcda
  root         /home/runner/work/zlib-ng/zlib-ng
  fname        /home/runner/work/zlib-ng/zlib-ng/insert_string_tpl.h
Parsing coverage data for file /home/runner/work/zlib-ng/zlib-ng/insert_string_tpl.h
```
I see nothing wrong here at all. Only possibility is the \n at the end of the file name, but it would be a bit strange if that was the error, no?

Relevant part of the codecov yaml below; it replaces that part of the path with nothing, so that it sees files according to the path in the repo. So that should still be correct.
```
fixes:
- '/home/runner/work/zlib-ng/zlib-ng/::'
```

**Comment by nmoinvaz** [comment]
I noticed in the CI list that the 2 Codecov instances no longer show.

**Comment by Dead2** [comment]
@nmoinvaz That is a problem, but unrelated to this PR. Codecov claims it was because of the buildkite build was failing and suggested config changes. I have applied the changes, but not sure whether it will work for old PRs.
In any case, those are missing from all PRs, not just this one. But this one is the only one without any uploaded coverage reports.

**Comment by Dead2** [comment]
Still nothing. Just to exclude some of the recent codecov problems, could you close this PR and make a new one?

**Comment by nmoinvaz** [comment]
It does report properly in my branch.
https://codecov.io/gh/nmoinvaz/zlib-ng/commit/baa58349fe038d70cc55a473ea36785c12d92b43/

**Comment by Dead2** [comment]
Then perhaps it is because it is a foreign repo and thus treated differently (hence our token workaround)

**Comment by nmoinvaz** [comment]
Yes, it is probably due to the token not being exposed to my PR. You could create a branch in zlib-ng and cherry-pick my changes and test it there?

**Comment by Dead2** [comment]
Sure, but we would still need to fix the workaround, so that we get coverage data for PRs like we do today.
As-is, this would be a regression in that regard.

**Comment by nmoinvaz** [comment]
It keeps saying there was an error in processing.. what does your codecov.yaml look like?
```
coverage:
  status:
    project:
      default:
        target: auto
        threshold: 1.0
fixes:
- '^build/home/runner/work/zlib-ng/zlib-ng/::'
- '^build/home/runner/work/zlib-ng/build/::'
ignore:
- ^usr/include/.*
- ^usr/lib/.*
- ^build/usr/include/.*
- ^build/usr/lib/.*
- ^lib/.*
- ^lib/.*/.*

```

**Comment by Dead2** [comment]
```
codecov:
  max_report_age: 72h
  notify:
    wait_for_ci: false
  require_ci_to_pass: false
comment:
  require_base: false
  require_head: false
coverage:
  status:
    project:
      default:
        threshold: 0.07
fixes:
- '/home/runner/work/zlib-ng/zlib-ng/::'
- '/home/runner/work/zlib-ng/build/::'
- '/build/home/runner/work/zlib-ng/zlib-ng/::'
- '/build/home/runner/work/zlib-ng/build/::'
ignore:
- usr/include/.*
- /usr/include/.*
- /build/usr/include/.*
- usr/lib/.*
- /usr/lib/.*
- /build/usr/lib/.*
- usr/lib64/.*
- /usr/lib64/.*
- /build/usr/lib64/.*
```

We really should add it as a file though.

**Comment by nmoinvaz** [comment]
Do you want me to add it as part of this PR?

**Comment by Dead2** [comment]
Sure, if you want to.
The fixes and ignore parts should be cleaned up first though, some of that stuff is old Travis stuff, but I have not wanted to mess with it as long as it worked correctly.

**Comment by nmoinvaz** [comment]
I can't figure out why the logs are showing `Token found by environment variables` even when I don't set environment variable.

**Comment by nmoinvaz** [comment]
Actually it looks like it is the `codecov/action` that is setting that environment variable for the NodeJS uploader.

**Comment by nmoinvaz** [comment]
@thomasrockhu any idea why this PR would not be showing up on the code coverage website? The reports generate properly and upload to codecov.io but fail to show on the website. 

They do however show for my repository, but not in the PR commit.
 
My PR for zlib-ng repo:
https://codecov.io/gh/zlib-ng/zlib-ng/commit/1b2851f3af49fa67fc34f2b4a24d7352cfb3bda4/
My branch for my repo:
https://codecov.io/gh/nmoinvaz/zlib-ng/commit/baa58349fe038d70cc55a473ea36785c12d92b43/

We are trying to move away from python uploader and to use the codecov GH action.

**Comment by nmoinvaz** [comment]
Thank you @thomasrockhu-codecov that seems to have fixed the issue. Not sure why though because I did check the timestamp being put in the XML and it looked right. Anyways, by turning that check off at least now there is one less point of failure. I ended up using `gcovr` instead of `pytest` because we needed to specify `gcov-executable` for some of our test runs.

**Comment by nmoinvaz** [comment]
@Dead2 I am seeing -5% coverage. I think these results are accurate. It could be because the benchmarks are not being run on every platform. I think after incorporating gtest (which has all the c++ tooling) we can turn on the benchmarks for more platforms.

**Comment by nmoinvaz** [comment]
Rebased.

**Comment by Dead2** [comment]
I really don't see any reason why this PR should change the coverage at all (other than the usual noise), this commit does not really make any changes to build, code or testing compared to the develop commit it is based on, so coverage should stay the same.

-5% is a pretty big change. Interestingly, most of the lost coverage is related to assert() or failure modes.
See test/infcover.c for example: https://app.codecov.io/gh/zlib-ng/zlib-ng/compare/1158/changes
I have no idea why that would be any different with this PR.

But I'll reluctantly approve this for merge.

**Review by Dead2 — APPROVED** [review]

**Comment by nmoinvaz** [comment]
@Dead2, I think it might have something to do with `gcovr` generating reports. I could try `--exclude-unreachable-branches` and see if that helps..

**Comment by nmoinvaz** [comment]
@Dead2 looks like that gave us +5% in coverage.

**Comment by Dead2** [comment]
@nmoinvaz I see, looking at the docs it seems there is a mismatch with how the statistics count different parts of the code. And `--exclude-unreachable-branches` seems to exclude a little more than the previous codecov-uploader does.
But from what I can see, there is nothing being excluded that we actually do care about.
`Exclude branch coverage from lines without useful source code (often, compiler-generated “dead” code).`, so things like initialization code, GCC attributes, closing curly braces, etc.

So, I think this new percentage is actually more useful to us than the old one was. 👍

File: .github/workflows/cmake.yml [hunk 1]
```diff
@@ -432,28 +432,34 @@ jobs:
...
-    - name: Upload coverage report
-      if: matrix.codecov && ( env.CODECOV_TOKEN_SECRET != '' || github.repository == 'zlib-ng/zlib-ng' )
+    - name: Generate coverage report
+      if: matrix.codecov
       shell: bash
       run: |
-        python -u -m pip install --user codecov
-        bash tools/codecov-upload.sh
+        python3 -u -m pip install --user gcovr
+        python3 -m gcovr --exclude-unreachable-branches --gcov-executable "${{ matrix.gcov-exec || 'gcov' }}" --root ${{ matrix.build-src-dir || '.' }} --xml --output coverage.xml --verbose
```
Comment (Dead2) [inline .github/workflows/cmake.yml]: The Github Actions VMs have 2 "cores", so perhaps get a little bit of parallelism going? 

I would think adding `-j3` would be a good setting (it is a lot of IO, so 3-4 is probably optimal) to speed up the coverage collection a little bit. It currently only takes about 5sec, but when we do that 43 times that could save a few seconds on how fast the full CI completes its queue.

File: .github/workflows/pigz.yml [hunk 2]
```diff
@@ -90,19 +85,23 @@ jobs:
...
 
-    - name: Upload coverage report
-      if: matrix.codecov && ( env.CODECOV_TOKEN_SECRET != '' || github.repository == 'zlib-ng/zlib-ng' )
-      shell: bash
+    - name: Generate coverage report
+      if: matrix.codecov
       run: |
-        bash tools/codecov-upload.sh
+        python3 -u -m pip install --user gcovr
+        python3 -m gcovr --exclude-unreachable-branches --gcov-executable "${{ matrix.gcov-exec || 'gcov' }}" --root . --xml --output coverage.xml --verbose
```
Comment (Dead2) [inline .github/workflows/pigz.yml]: Same here

File: .github/workflows/cmake.yml (hunk 1 above)
Comment (nmoinvaz) [reply .github/workflows/cmake.yml]: Done.

File: .github/workflows/pigz.yml (hunk 2 above)
Comment (nmoinvaz) [reply .github/workflows/pigz.yml]: Done.

**Pull request merged**